neural_lam/weather_dataset.py

```diff
@@ -1,262 +1,364 @@
+                elapsed_forecast_duration=slice(
+                    n_timesteps_offset, n_steps + n_timesteps_offset
+                ),
+            )
+            # create a new time dimension so that the produced sample has a
+            # `time` dimension, similarly to the analysis only data
+            da["time"] = da.analysis_time + da.elapsed_forecast_duration
+            da = da.swap_dims({"elapsed_forecast_duration": "time"})
+        else:
+            # only `time` dimension for analysis only data
+            da = da.isel(
+                time=slice(idx + n_timesteps_offset, idx + n_steps + n_timesteps_offset)
+            )
+        return da
 
     def __getitem__(self, idx):
-        # === Sample ===
-        sample_name = self.sample_names[idx]
-        sample_path = os.path.join(
-            self.sample_dir_path, f"nwp_{sample_name}.npy"
-        )
-        try:
-            full_sample = torch.tensor(
-                np.load(sample_path), dtype=torch.float32
-            )  # (N_t', dim_y, dim_x, d_features')
-        except ValueError:
-            print(f"Failed to load {sample_path}")
-
-        # Only use every ss_step:th time step, sample which of ss_step
-        # possible such time series
-        if self.random_subsample:
-            subsample_index = torch.randint(0, self.subsample_step, ()).item()
+        """Return a single training sample, which consists of the initial states, target
+        states, forcing and batch times.
+
+        The implementation currently uses xarray.DataArray objects for the
+        normalisation so that we can make us of xarray's broadcasting
+        capabilities. This makes it possible to normalise with both global
+        means, but also for example where a grid-point mean has been computed.
+        This code will have to be replace if normalisation is to be done on the
+        GPU to handle different shapes of the normalisation.
+
+        Parameters
+        ----------
+        idx : int
+            The index of the sample to return, this will refer to the time of
+            the initial state.
+
+        Returns
+        -------
+        init_states : TrainingSample
+            A training sample object containing the initial states, target
+            states, forcing and batch times. The batch times are the times of
+            the target steps.
+
+        """
+        # handling ensemble data
+        if self.datastore.is_ensemble:
+            # for the now the strategy is to simply select a random ensemble member
+            # XXX: this could be changed to include all ensemble members by
+            # splitting `idx` into two parts, one for the analysis time and one
+            # for the ensemble member and then increasing self.__len__ to
+            # include all ensemble members
+            i_ensemble = np.random.randint(self.da_state.ensemble_member.size)
+            da_state = self.da_state.isel(ensemble_member=i_ensemble)
```

Comment:
Note that training on ensemble forecasts (and on ensemble reanalysis, not sure if this would supports that?) is something quite orthogonal to running probabilistic models that produce ensemble forecasts. These things could be more tightly coupled, but the way I have done this so far there is no connection between training on ensembles and forecasting ensembles. So I don't think this belongs together with #62.

Reply:
I'll resolve this since we'll continue the ensemble discussion elsewhere :)